src/Microsoft.AspNetCore.Routing/EndpointFeature.cs

```diff
@@ -54,7 +54,20 @@ RouteData IRoutingFeature.RouteData
             {
                 if (_routeData == null)
                 {
-                    _routeData = new RouteData(_values);
+                    _routeData = _values == null ? new RouteData() : new RouteData(_values);
+
+                    // Note: DataTokens won't update if someone else overwrites the Endpoint
+                    // after route values has been set. This seems find since endpoints are a new
+                    // feature and DataTokens are for back-compat.
```

Comment:
I didn't have a better idea than this. RouteValues is really commonly read/write. Data tokens isn't frequently written to in application code.